### PR TITLE
[go_router] Add `buildPageWithState`

### PR DESCRIPTION
**Pull request by ValentinVignal**
This should close https://github.com/flutter/flutter/issues/109440

*edited by @chunhtai*

should close https://github.com/flutter/flutter/issues/103985

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by ValentinVignal** [comment]
This is a _"dummy"_ implementation because I don't know what you would prefer.

Do you want me to change the existing `buildPage` method? But it would be a breaking change.
Should I create a new method as I did? If yes, do you prefer another name than `buildPageWithState` ?

For the tests, I guess their implementation/logic would change depending on what you tell me to do. I'll write them once I know what I should implement

File: packages/go_router/lib/src/route_data.dart [hunk 1]
```diff
@@ -40,6 +40,10 @@ abstract class GoRouteData {
   /// [Page] implementation to be used with the results of [build].
   Page<void> buildPage(BuildContext context) => const NoOpPage();
 
+  /// A page builder for that expose the [state].
+  Page<void> buildPageWithState(BuildContext context, GoRouterState state) =>
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:44]: This should return buildPage by default

File: packages/go_router/lib/src/route_data.dart [hunk 2]
```diff
@@ -82,8 +86,14 @@ abstract class GoRouteData {
     Widget builder(BuildContext context, GoRouterState state) =>
         factoryImpl(state).build(context);
 
-    Page<void> pageBuilder(BuildContext context, GoRouterState state) =>
-        factoryImpl(state).buildPage(context);
+    Page<void> pageBuilder(BuildContext context, GoRouterState state) {
+      final T routeData = factoryImpl(state);
+      final Page<void> page = routeData.buildPageWithState(context, state);
+      if (page is! NoOpPage) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:92]: you can skip this check and just return the page if you make buildPageWithState return buildPage by default

**Review by chunhtai — COMMENTED** [review]
adding a new API to avoid breaking change sounds good to me

**Comment by ValentinVignal** [comment]
Thank you @chunhtai 
Should I deprecate `buildPage` ?

File: packages/go_router/lib/src/route_data.dart (hunk 1 above)
Comment (ValentinVignal) [reply packages/go_router/lib/src/route_data.dart:56]: Nice, I did it in https://github.com/flutter/packages/pull/2444/commits/b67e5f8fa369ad12631b072f325416a5881136bc

File: packages/go_router/lib/src/route_data.dart (hunk 2 above)
Comment (ValentinVignal) [reply packages/go_router/lib/src/route_data.dart]: 👍 I've applied your suggestion in https://github.com/flutter/packages/pull/2444/commits/b67e5f8fa369ad12631b072f325416a5881136bc

**Comment by ValentinVignal** [comment]
I deprecated `buildPage` and the analyzer is failing because of that. What should I do about it? Add `// ignore: ...` ?

**Comment by chunhtai** [comment]
> I deprecated `buildPage` and the analyzer is failing because of that. What should I do about it? Add `// ignore: ...` ?

for the reference in doc string, please update it to use buildPageWithState. For the code that uses it, you can add ignore if it is unavoidable; otherwise, they should use buildPageWithState.

You may need to add ignore to the go_router_builder first

**Comment by ValentinVignal** [comment]
@chunhtai I replaced the references in the doc strings and ignore the call to `builPage` in https://github.com/flutter/packages/pull/2444/commits/32b94eb988ab8bdd1d2efb6a0815cf29ab8c57f7

File: packages/go_router/lib/src/route_data.dart [hunk 3]
```diff
@@ -32,17 +34,35 @@ abstract class GoRouteData {
...
   ///
-  /// Subclasses must override one of [build], [buildPage], or [redirect].
+  /// Subclasses must override one of [build], [buildPageWithState] or
+  /// [redirect].
   ///
   /// Corresponds to [GoRoute.pageBuilder].
   ///
   /// By default, returns a [Page] instance that is ignored, causing a default
   /// [Page] implementation to be used with the results of [build].
+  @Deprecated('Use buildPageWithState instead.')
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:44]: This should follow the same format as flutter repo
for example
```
  @Deprecated(
    'This method has been deprecated in favor of _waitForCondition. '
    'This feature was deprecated after v1.9.3.'
  )
```

File: packages/go_router/lib/src/route_data.dart [hunk 4]
```diff
@@ -32,17 +34,35 @@ abstract class GoRouteData {
...
   Page<void> buildPage(BuildContext context) => const NoOpPage();
 
+  /// A page builder for this route that expose the [state].
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:47]:
```suggestion
  /// A page builder for this route with [GoRouterState].
```

**Review by chunhtai — COMMENTED** [review]
Code Looks good to me, but the go_router_builder may need to be marked //ignore first.

**Comment by ValentinVignal** [comment]
>  the go_router_builder may need to be marked //ignore first.

I'm sorry, I don't understand what you mean by that

**Comment by ValentinVignal** [comment]
> > the go_router_builder may need to be marked //ignore first.
> 
> I'm sorry, I don't understand what you mean by that

@chunhtai is https://github.com/flutter/packages/pull/2444/commits/d5f9c9261eee7860739ae93d44649e3596aba9e3 what you meant ?

**Comment by chunhtai** [comment]
> > > the go_router_builder may need to be marked //ignore first.
> > 
> > 
> > I'm sorry, I don't understand what you mean by that
> 
> @chunhtai is [d5f9c92](https://github.com/flutter/packages/commit/d5f9c9261eee7860739ae93d44649e3596aba9e3) what you meant ?

yes.

**Review by loic-sharma — APPROVED** [review]

**Comment by chunhtai** [comment]
@ValentinVignal the ignore in go_router_builder should be in a separate pr that needs to go in first.

**Comment by ValentinVignal** [comment]
> @ValentinVignal the ignore in go_router_builder should be in a separate pr that needs to go in first.

Ok, I've create this PR for the ignore in go_router_builder https://github.com/flutter/packages/pull/2504

File: packages/go_router_builder/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+## 1.0.11
```
Comment (chunhtai) [inline packages/go_router_builder/CHANGELOG.md:1]: This should be in a separate pr. We should avoid touching different packages in same pr, it will make roll back easier if needed to.

**Review by chunhtai — COMMENTED** [review]
code looks good except for modification on go_router_builder

**Comment by ValentinVignal** [comment]
@chunhtai https://github.com/flutter/packages/pull/2504 got merged, so I guess this PR can be re-reviewed

File: packages/go_router_builder/CHANGELOG.md (hunk 5 above)
Comment (ValentinVignal) [reply packages/go_router_builder/CHANGELOG.md]: Faire enough, I've created this PR https://github.com/flutter/packages/pull/2535 on `go_router_builder` and discarded the changes on `go_router_builder` in this PR in  https://github.com/flutter/packages/pull/2444/commits/4aa91506f5f1196e0728a28ef6be91cf47b8bd3c and https://github.com/flutter/packages/pull/2444/commits/66b94dfdc11d7eddffaba6a596db6358f70a415c

File: packages/go_router/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,8 @@
+## 4.4.0
+
+- Add `buildPageWithState` to `GoRouteData`.
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:3]:
```suggestion
- Adds `buildPageWithState` to `GoRouteData`.
```

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by ValentinVignal** [comment]
@chunhtai Can I do a merge or does it have to be a rebase?

**Comment by chunhtai** [comment]
rebase will be cleaner and that is what I usually do.

**Comment by ValentinVignal** [comment]
@chunhtai I merged `main` in my branch in https://github.com/flutter/packages/pull/2444/commits/3802759ddfd06eef5842c7ad2fd89d20b8a18a82 
And I think I managed to merge without causing commits hell. Tell me if that is good for you or not
